tree/treeplayer/src/TTreeProcessorMT.cxx

```diff
@@ -167,13 +168,30 @@ static ClustersAndEntries MakeClusters(const std::vector<std::string> &treeNames
       std::vector<EntryCluster> clusters;
       while ((start = clusterIter()) < entries) {
          end = clusterIter.GetNextEntry();
-         // Add the current file's offset to start and end to make them (chain) global
-         clusters.emplace_back(EntryCluster{start + offset, end + offset});
+         // Currently, if a user specified a range, the clusters will be only globally obtained
+         // Hence, the guarantee that currentStart and currentEnd are always as expected
+         // That are: the cluster start must be at least startEntry, and at most the endEntry
+         // The current file's offset to start and end is added to make them (chain) global
+         const auto currentStart = std::max(start + offset, startEntry);
+         const auto currentEnd = std::min(end + offset, endEntry);
+         // This is not satified if the desired start is larger than the last entry of some cluster
+         // In this case, this cluster is not going to be processes further
+         if (currentStart < currentEnd)
+            clusters.emplace_back(EntryCluster{currentStart, currentEnd});
+         if (currentEnd == endEntry) // if the desired end is reached, stop reading further
+            break;
       }
-      offset += entries;
+      offset += entries; // consistently keep track of the total number of entries
       clustersPerFile.emplace_back(std::move(clusters));
+      // The clusters might be empty, but their entries are needed to properly handle the global range
+      // This can happen if a cluster had last entry smaller than the desired start. The position of the
+      // desired start must remain globally the same.
       entriesPerFile.emplace_back(entries);
    }
+   if (startEntry >= offset)
+      throw std::logic_error(std::string("A range of entries was passed in the creation of the RDataFrame, ") +
+                             "but the starting entry is larger than the total number of entries (" + offset +
+                             ") in the dataset.");
```

Comment:
this message should not mention RDF as we are in TTreeProcessorMT.cxx -- but maybe we should just take this as a pre-condition in `MakeClusters` and ask the callers to guarantee this: I'm not sure it's `MakeCluster`'s job to check, this function can correctly do its job even if `startEntry >= offset`.